flang/lib/Lower/ConvertProcedureDesignator.cpp

```diff
@@ -125,3 +130,15 @@ hlfir::EntityWithAttributes Fortran::lower::convertProcedureDesignatorToHLFIR(
       [funcAddr](const auto &) { return funcAddr; });
   return hlfir::EntityWithAttributes{res};
 }
+
+mlir::Value Fortran::lower::convertProcedureDesignatorToAddress(
+    Fortran::lower::AbstractConverter &converter, mlir::Location loc,
+    mlir::Type boxType, Fortran::lower::StatementContext &stmtCtx,
+    const Fortran::semantics::Symbol *sym) {
+  Fortran::lower::SymMap globalOpSymMap;
+  Fortran::evaluate::ProcedureDesignator proc(*sym);
+  auto procVal{Fortran::lower::convertProcedureDesignatorToHLFIR(
+      loc, converter, proc, globalOpSymMap, stmtCtx)};
+  return fir::getBase(Fortran::lower::convertToAddress(
+      loc, converter, procVal, stmtCtx, procVal.getType()));
```

Comment:
Since `boxType` is passed, I think it would be safer to do a builder.convert here just in case the target symbol and procedure pointer interface to not lower to the exact same MLIR function type.

Reply:
I am afraid we may hit verifier errors in some cases (I need to find a convincing example). The HasValueOp operand must have the same type as the fir.globalOp.

- The fir.globalOp takes its type from the procedure pointer interface
- The type of `procVal` passed later to HasValueOp is from sym whose type is from the interface of the intial target.

If for any reasons the interfaces of the procedure pointer and its initial targets are compatible "green light from semantics" but do not lower to the exact same MLIR signature, this will be a bug here. I am not convinced we should rely on the MLIR types always being the same, even though I do not have an immediate Fortran example that I can come up with.

Reply:
But it is also OK to do the convert inside just before the HasValueOp if needed instead of passing boxType as an argument and doing that here.

Reply:
Agreed. I will add `auto castTo{builder.createConvert(loc, symTy, box)};` to convert to the declared type of the procedure pointer.